src/Physics/BeamHNL/HNLDecayVolume.h

```diff
@@ -0,0 +1,171 @@
+//____________________________________________________________________________
+/*!
+
+\class     genie::HNL::HNLDecayVolume
+
+\brief     Utilities for extracting the position of the HNL decay vertex
+           *** given production vertex and momentum ***
+
+\author    John Plows <komninos-john.plows \at physics.ox.ac.uk>
+          
+\created   March 31st, 2022
+
+\cpright   Copyright (c) 2003-2022, The GENIE Collaboration
+           For the full text of the license visit http://copyright.genie-mc.org          
+*/
+//____________________________________________________________________________
+
+#ifndef _HNL_DECAY_VOLUME_H_
+#define _HNL_DECAY_VOLUME_H_
+
+#include <cmath>
+#include <cassert>
+
+#include <TVector3.h>
+//#ifdef __GENIE_GEOM_DRIVERS_ENABLED__ // why do we crash with this guard on?
+#include <TGeoManager.h>
+#include <TGeoVolume.h>
+#include <TGeoNode.h>
+#include <TGeoBBox.h>
+//#endif // #ifdef __GENIE_GEOM_DRIVERS_ENABLED__
+
+//#include "Framework/Conventions/Constants.h"
+#include "Framework/Conventions/Controls.h"
+#include "Framework/Conventions/Units.h"
+#include "Framework/GHEP/GHepParticle.h"
+#include "Framework/GHEP/GHepRecord.h"
+#include "Framework/Messenger/Messenger.h"
+#include "Framework/Numerical/RandomGen.h"
+#include "Framework/Utils/UnitUtils.h"
+#include "Framework/Utils/PrintUtils.h"
+
+#include "Physics/BeamHNL/HNLDecayer.h"
+
```

Comment:
`DecayVolume` probably deserves a better name. It acts on the EventRecord by calculating where the HNL would have entered the user-input geometry, where it would have exited it, and applying an exponentially decaying pdf assigns a decay vertex (`event->Vertex()`) somewhere along that trajectory.
One cannot simply distribute the decay vertices in the input decay volume at the App stage, because of the POT calculation: the HNL was made at specific coordinates (given by user-input beamline simulation, `-f`) and travels along a line calculated in `DecayVolume`. This impacts the probability a) of the HNL surviving until it enters the Volume and b) of the HNL decaying before it can exit the Volume. 
That said, the `DecayVolume` now implements the `EventRecordVisitorI` interface without any additional public methods; all information that was handed by public calls is now written inside the EventRecord and so eliminates the need for the extra public stuff apart from `ProcessEventRecord`.

Reply:
By "acts on the EventRecord", 2 things are meant: 1) the Vertex() is assigned, **and 2) the Weight() is updated**. Without this update, the impact of detector location / size on sensitivity is lost.

Reply:
HNLVertexGenerator? There is a VertexGenerator in other generators too. A good name is important, through the main concern was whether this was indeed a event record visitor (that needs only the standard event record visitor interface) or something else that was better split over multiple classes. If you were able to hide all the functions that were previously public, that's a good sign!

Reply:
Name has been changed to HNLVertexGenerator.